app/uk/gov/hmrc/agentsubscription/controllers/SubscriptionJourneyController.scala

```diff
@@ -0,0 +1,100 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.agentsubscription.controllers
+
+import java.time.{ LocalDateTime, ZoneId, ZoneOffset }
+import java.util.UUID
+
+import com.google.inject.Inject
+import com.kenshoo.play.metrics.Metrics
+import play.api.libs.json.Json.toJson
+import play.api.libs.json.{ JsError, JsSuccess, JsValue }
+import play.api.mvc.{ Action, AnyContent, Result }
+import uk.gov.hmrc.agentmtdidentifiers.model.Utr
+import uk.gov.hmrc.agentsubscription.auth.AuthActions
+import uk.gov.hmrc.agentsubscription.connectors.MicroserviceAuthConnector
+import uk.gov.hmrc.agentsubscription.model.InternalId
+import uk.gov.hmrc.agentsubscription.model.subscriptionJourney.SubscriptionJourneyRecord
+import uk.gov.hmrc.agentsubscription.repository.SubscriptionJourneyRepository
+import uk.gov.hmrc.play.bootstrap.controller.BaseController
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+class SubscriptionJourneyController @Inject() (implicit
+  metrics: Metrics,
+  microserviceAuthConnector: MicroserviceAuthConnector,
+  subscriptionJourneyRepository: SubscriptionJourneyRepository,
+  ec: ExecutionContext)
+  extends AuthActions(metrics, microserviceAuthConnector) with BaseController {
+
+  private def localWithJsonBody(f: SubscriptionJourneyRecord => Future[Result], request: JsValue): Future[Result] =
+    Try(request.validate[SubscriptionJourneyRecord]) match {
+      case Success(JsSuccess(payload, _)) => f(payload)
+      case Success(JsError(errs)) => Future successful BadRequest(s"Invalid payload: $errs")
+      case Failure(e) => Future successful BadRequest(s"could not parse body due to ${e.getMessage}")
+    }
+
+  def findByPrimaryId(internalId: InternalId): Action[AnyContent] = Action.async { implicit request =>
+    subscriptionJourneyRepository.findByPrimaryId(internalId).map {
+      case Some(record) => Ok(toJson(record))
+      case None => NoContent
+    }
+  }
+
+  def findByMappedId(internalId: InternalId): Action[AnyContent] = Action.async { implicit request =>
+    subscriptionJourneyRepository.findByMappedId(internalId).map {
+      case Some(record) => Ok(toJson(record))
+      case None => NoContent
+    }
+  }
+
+  def findByUtr(utr: Utr): Action[AnyContent] = Action.async { implicit request =>
+    subscriptionJourneyRepository.findByUtr(utr).map {
+      case Some(record) => Ok(toJson(record))
+      case None => NoContent
```

Comment:
if the entity doesn't exists, why do we return NoContent instead of NotFound ?

Reply:
Well, this is a subtle thing - 404 actually means: I can't find the URL you asked for.  It's what you would return if the URL was not found.  The URL is correct; but there is no resource there.  Also, on a more practical level, the HMRC http client library treats *all* 4xx responses as exceptions.  No record found is not an exception/error - it is normal behaviour.  Therefore, no content is a reasonable response here - there was no error, we just didn't find anything.  Actually, there is also an argument to return 200 (with empty body) rather than 204, but I think it's to distinguish by response code rather than just body content.

Reply:
if the findBy url is correct REST Url, and if there is no resource, then it should be 404 , 
yes 404 is not an exception( its an issue with hmrc library), i remember using `RawHttpReads` in `help-to-save` project to overcome that issue 
and if i remember correctly 204/200 codes are for POST/PUT/DELETE, but if its a GET then its appropriate to return 404 (?)